src/Enable.Extensions.Queuing.RabbitMQ/Internal/ConnectionManagement/RabbitMQConnectionManager.cs

```diff
@@ -0,0 +1,157 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Enable.Extensions.Queuing.RabbitMQ.Internal.ConnectionManagement;
+using RabbitMQ.Client;
+using RabbitMQ.Client.Exceptions;
+
+namespace Enable.Extensions.Queuing.RabbitMQ.Internal
+{
+    internal class RabbitMQConnectionManager
+    {
+        private static readonly List<(int hash, RabbitMQManagedConnection managedConnection)>
+            _connections = new List<(int hash, RabbitMQManagedConnection managedConnection)>();
+
+        private static readonly object _lock = new object();
+
+        private readonly int _connectionHash;
+        private readonly IConnectionFactory _connectionFactory;
+        private readonly RabbitMQQueueClientFactoryOptions _options;
+
+        public RabbitMQConnectionManager(RabbitMQQueueClientFactoryOptions options)
+        {
+            _options = options;
+
+            _connectionFactory = new ConnectionFactory
+            {
+                HostName = options.HostName,
+                Port = options.Port,
+                VirtualHost = options.VirtualHost,
+                UserName = options.UserName,
+                Password = options.Password,
+                AutomaticRecoveryEnabled = options.AutomaticRecoveryEnabled,
+                NetworkRecoveryInterval = options.NetworkRecoveryInterval
+            };
+
+            _connectionHash = new
+            {
+                _options.HostName,
+                _options.Port,
+                _options.VirtualHost,
+                _options.UserName,
+                _options.Password,
+                _options.AutomaticRecoveryEnabled,
+                _options.NetworkRecoveryInterval,
+            }.GetHashCode();
+        }
+
+        /// <summary>
+        /// Provides a new RabbitMQ channel, session and model. The underlying connection is managed,
+        /// ensuring a connection is reused if possible.
+        /// </summary>
+        public RabbitMQManagedChannel CreateChannel()
+        {
+            RabbitMQManagedConnection managedConnection = null;
+            IModel channel = null;
+
+            managedConnection = GetConnection();
+
+            try
+            {
+                channel = managedConnection.Connection.CreateModel();
+            }
+            catch (Exception ex)
+            {
+                if (ex is ChannelAllocationException || ex is AlreadyClosedException)
```

Comment:
Perhaps it would be nicer if we used [C# Exception filtering](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-6#exception-filters) here?

Reply:
Thanks for the suggestion! I've updated the code to use exception filtering.